### PR TITLE
Only redefine association autosave callback if it isn't defined by the current class

### PR DESCRIPTION
**Pull request by al2o3cr**
Addresses #17015 and #16494. Reverts the majority of the changes from #15358 but retains the test coverage.

TL;DR on the discussion in #17015: the fix in #15358 ensured that subclasses that redeclare associations get newly-defined autosave callbacks. Unfortunately, enabling `accepts_nested_attributes_for` can cause this redefinition within a single class, breaking the order of callbacks and causing records to be saved in the wrong order.

This PR reinstates the method-definition check removed in #15358 but restricts it to only match definitions on the current class rather than inherited ones.

Several other changes are included, mostly for symmetry and to prevent additional similar bugs:
- the definition in the `reflection.has_one?` branch now uses `define_non_cyclic_method` to match the other branches of the conditional. It could be argued that two models who both `has_one` each other are probably incorrect DB design, but that still shouldn't produce infinite recursion.
- similarly, the `method_defined?` check in the validation conditional (line 208) has been changed to match the check for `save_method` on line 182. I don't have a test case for this, but it would likely manifest in the same "redefinition in subclasses" sorts of situations that #15358 addressed.

I've included the test written by @pranas and mentioned in https://github.com/rails/rails/issues/17015#issuecomment-67736744 as a separate commit as it accurately reflects the code's authorship. Let me know if I should squash, etc.

4.2.0 is also affected by this bug. Happy to backport this when it gets merged.

/cc @tenderlove @sgrif

### CODE REVIEW

**Comment by robin850** [comment]
As the changelog entry, what do you think about something along the lines of:

> *   Avoid redefining the association auto-save callbacks if they are already
>     defined by the current class. This was mostly a problem dealing with
>     associations that were linked with `accepts_nested_attributes_for`.
>
>     Fixes #17015 and #16494.
>
>    *Matt Jones* and *Pranas Kiziela*

Thanks for the patch by the way! :+1:

**Comment by al2o3cr** [comment]
Update on this: the call to `add_autosave_association_callbacks` isn't in the `accepts_nested_attributes_for` machinery on master. The test from @pranas no longer fails.

I suspect this is still potentially triggerable by redeclaring the join table association in a subclass, as the callbacks will get added a second time. Will investigate further.

**Comment by tekin** [comment]
@al2o3cr did you get any further with this? I'm happy to take a look and pick up the mantle if it's fallen off your radar.

**Comment by al2o3cr** [comment]
@tekin - the general issue still exists (redefining associations in subclasses can reorder callbacks) but the specific one in `accepts_nested_attributes_for` was removed by not re-addding the callbacks.

Long-term, I believe the solution is to explicitly construct the dependency graph of records-to-be-saved instead of relying on callback order. I've done some toy models that indicate that this may be possible with `tsort`, but nothing production-ready yet.

I'll definitely take a look at your models, though - producing test cases to trigger this bug is tricky, so it's handy to have a real-world example. It's also worth checking if your models produce this behavior on 5.0 / master.

**Comment by tekin** [comment]
@al2o3cr thanks for the update. You can find the models in question over at [alphagov/whitehall](https://github.com/alphagov/whitehall). Specifically the `Classification`/`Topic` STI with the `organisation_classification` join model. Here's where [the commit](https://github.com/alphagov/whitehall/commit/e2509c42f01da53d82cdad5fd7e6981745b69342) that specifically works around the bug.

**Comment by rbr** [comment]
Any updates on this?
I'm running into the same issue as in #16494, which appears is the same this PR addresses. 

A work-around is the there mentioned calling the `accepts_nested_attributes_for` before the `has_many :through`, but that doesn't really feel like a clean solution.

**Comment by kamipo** [comment]
This isn't failed on master. Closing.
